### PR TITLE
Kernel+Services: Introduce the DeviceMapper - a barebones hot-plug daemon handler

### PR DESCRIPTION
**Pull request by supercomputer7**
We already have AHCI support for hotplugging SATA devices. To be able to leverage this so far we needed to manually add a device node (or more than 1) in `/dev`. This is a preparation before USB HID could be added safely, as we will detect them in asynchronously way, so ideally `WindowServer` will get a notification about USB HID device being attached from `SystemServer`, based on this code in this PR, which will be enhanced to propagate notifications in a future PR :)
Here's a discussion with @FireFox317 why this is necessary - https://discord.com/channels/830522505605283862/897113928168013855/1101937994828628109

This currently blocks the following work on these PRs:
- Kernel: Implement support for PS2 scan code set 2 #18342 
- Kernel: Introduce basic USB HID support (keyboard and mouse) #18673 
- Kernel/SerialIO: Introduce basic PS2 core for device detection #18792 

Merging this will immediately unblock work on PR #18342.

### CODE REVIEW

**Comment by supercomputer7** [comment]
The next-next PR to be in the queue, as I said before, is #12885.

**Comment by supercomputer7** [comment]
CI was red, so let's see if it's now green. Anyway, I wrote the commit messages, so reviews on this are appreciated as always!

File: Userland/Services/SystemServer/DeviceEventLoop.cpp [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
+{
+    auto base_string = TRY(String::from_utf8(""sv));
+    do {
+        auto current_char = offset_character_with_number('a', allocation_index % 26);
+        base_string = TRY(String::formatted("{:c}{}", current_char, base_string));
+        allocation_index = allocation_index / 26;
+    } while (allocation_index > 0);
+    return base_string;
+}
```
Comment (BertalanD) [inline Userland/Services/SystemServer/DeviceEventLoop.cpp:106]: nit: This could be written a bit terser

```suggestion
static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
{
    auto base_string = TRY(String::from_utf8(""sv));
    do {
        base_string = TRY(String::formatted("{:c}{}", 'a' + (allocation_index % 26), base_string));
        allocation_index = allocation_index / 26;
    } while (allocation_index > 0);
    return base_string;
}
```

File: Userland/Services/SystemServer/main.cpp [hunk 2]
```diff
@@ -392,49 +123,46 @@ static ErrorOr<void> prepare_bare_minimum_devtmpfs_directory_structure()
     TRY(Core::System::mkdir("/dev/gpu"sv, 0755));
     TRY(Core::System::mkdir("/dev/pts"sv, 0755));
     TRY(Core::System::mount(-1, "/dev/pts"sv, "devpts"sv, 0));
+
+    mode_t old_mask = umask(0);
+    printf("Changing umask %#o\n", old_mask);
```
Comment (BertalanD) [inline Userland/Services/SystemServer/main.cpp:128]: nit: I know that this is copy-pasta, but since you're already touching it:
```suggestion
```

(I don't think this message ever shows up as the stdout of SystemServer is not hooked up to anything)

File: Userland/Services/SystemServer/main.cpp [hunk 3]
```diff
@@ -520,6 +210,82 @@ static ErrorOr<void> create_tmp_semaphore_directory()
...
+    for (auto& service : g_services) {
+        if (auto result = service->activate(); result.is_error())
+            dbgln("{}: {}", service->name(), result.release_error());
+    }
+    return {};
+}
+
+static ErrorOr<void> activate_base_services_based_on_system_mode()
+{
+    pid_t pid = TRY(Core::System::fork());
```
Comment (BertalanD) [inline Userland/Services/SystemServer/main.cpp:252]: I don't see a fork in the original code. Why is this needed?

File: Userland/Services/SystemServer/main.cpp [hunk 4]
```diff
@@ -520,6 +210,82 @@ static ErrorOr<void> create_tmp_semaphore_directory()
...
+                found_gpu_device = true;
+                break;
+            }
+            sleep(1);
+        }
+        if (!found_gpu_device) {
+            dbgln("WARNING: No device nodes at /dev/gpu/ directory after 10 seconds. This is probably a sign of disabled graphics functionality.");
+            dbgln("To cope with this, graphical mode will not be enabled.");
+            g_system_mode = "text";
+        }
```
Comment (BertalanD) [inline Userland/Services/SystemServer/main.cpp:268]: Looks like we now hit this timeout on CI. Missing a check for `g_system_mode == "graphical"sv` before entering the loop?

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+    for (auto& family : m_device_node_families) {
+        if (family->major_number() == major_number && type == family->type() && is_in_family_minor_number_range(*family, minor_number))
+            return family.ptr();
+    }
+    return nullptr;
+}
+
+ErrorOr<DeviceNodeFamily*> DeviceEventLoop::find_or_register_new_device_node_family(device_node_match const& match, bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    if (auto* family = find_device_node_family(block_device, major_number, minor_number); !!family)
```
Comment (kleinesfilmroellchen) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:72]: (Sidenote for using references instead of pointers: this will also become more readable; instead of the weird `!!` you'll get `family.has_value()`)

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    TRY(m_device_node_families.try_append(move(node)));
+
+    return m_device_node_families.last().ptr();
+}
+
+static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
+{
+    auto base_string = TRY(String::from_utf8(""sv));
+    do {
+        base_string = TRY(String::formatted("{:c}{}", 'a' + (allocation_index % 26), base_string));
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Userland/Services/DeviceMapper/DeviceEventLoop.cpp:92]: You read it wrong - it will add a new letter one after one (so there could be a device file in `/dev` called `/dev/hdaab` :)

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+
+    auto path = TRY(String::from_utf8(path_pattern));
+    if (match.path_pattern.contains("%d"sv)) {
+        auto replacement = TRY(build_suffix_with_numbers(allocated_suffix_index));
+        path = TRY(path.replace("%d"sv, replacement, ReplaceMode::All));
+    }
+    if (match.path_pattern.contains("%c"sv)) {
+        auto replacement = TRY(build_suffix_with_letters(allocated_suffix_index));
+        path = TRY(path.replace("%c"sv, replacement, ReplaceMode::All));
+    }
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Userland/Services/DeviceMapper/DeviceEventLoop.cpp:140]: It's basically the same like what `printf()` does, so I can't really understand what is the problem. Maybe adding a comment will solve this?

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+ErrorOr<void> DeviceEventLoop::unregister_device(bool block_device, MajorNumber major_number, MinorNumber minor_number)
+{
+    if (!device_node_family_to_match_type(block_device, major_number, minor_number))
+        return {};
+    auto* family = find_device_node_family(block_device, major_number, minor_number);
+    if (!family) {
+        // FIXME: Is this an actual bug?
+        return Error::from_errno(ENODEV);
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Userland/Services/DeviceMapper/DeviceEventLoop.cpp:168]: Hmm, what if we try to remove a device node that wasn't even registered?

A scenario where this could happen is if we run the `DeviceMapper` program after the user generated a device node and after the hotplug event was discarded because of a previous instance that was either killed or crashed...

File: Userland/Services/DeviceMapper/DeviceEventLoop.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    };
+
+    DeviceEventLoop(NonnullOwnPtr<Core::File>);
+    virtual ~DeviceEventLoop() = default;
+
+    void drain_events_from_devctl();
+
+private:
+    DeviceNodeFamily* find_device_node_family(bool block_device, MajorNumber major_number, MinorNumber minor_number) const;
+    ErrorOr<DeviceNodeFamily*> find_or_register_new_device_node_family(device_node_match const& match, bool block_device, MajorNumber major_number, MinorNumber minor_number);
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Userland/Services/DeviceMapper/DeviceEventLoop.h]: Since we cannot have `ErrorOr<DeviceNodeFamily&>` I decided that `DeviceNodeFamily` should be Ref-Counted.
Really this makes sense, because in the future when other servers and services will query the `DeviceMapper` service, we will need to iterate over the list of `DeviceNodeFamily`s, and we definitely don't want to get into a situation where such object is removed while we use it.

File: Kernel/Devices/DeviceManagement.cpp [hunk 10]
```diff
@@ -52,14 +52,6 @@ Device* DeviceManagement::get_device(MajorNumber major, MinorNumber minor)
     });
```
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/Devices/DeviceManagement.cpp:52]: But `select()` actually needs proper blocking support.
See this:
https://github.com/SerenityOS/serenity/blob/bfd6deed1e860ce9516c7ca1613150a2c2c75f75/Kernel/FileSystem/File.h#L54-L59
This is exactly what I change - I implement `can_read` properly now.

>  can't think of any long-blocking device file at the moment that isn't a socket.

This is not correct, there are other device files that can block. See this:
https://github.com/SerenityOS/serenity/blob/bfd6deed1e860ce9516c7ca1613150a2c2c75f75/Kernel/Devices/SerialDevice.cpp#L24-L27

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Services/DeviceMapper/DeviceEventLoop.cpp:92]: Ah, I see now.

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Services/DeviceMapper/DeviceEventLoop.cpp:139]: However, especially %c is not really comparable to printf since it can print more than one character. The documentation must then be with the path pattern variable.

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Services/DeviceMapper/DeviceEventLoop.cpp:167]: Restarts is a valid scenario, but I feel like in that case we should rather be able to recover most of our internal device library manually. But I guess it’s fine for now.

File: Kernel/Devices/DeviceManagement.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (supercomputer7) [reply Kernel/Devices/DeviceManagement.cpp:52]: @kleinesfilmroellchen agreed on resolving this - https://discord.com/channels/830522505605283862/859531354219872266/1126228397400531035

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+{
+    mode_t old_mask = umask(0);
+    TRY(Core::System::create_char_device("/dev/kcov"sv, 0666, 30, 0));
+    umask(old_mask);
+    return {};
+}
+
+ErrorOr<void> DeviceEventLoop::read_one_or_eof(DeviceEvent& event)
+{
+    auto const read_buf = TRY(m_devctl_file->read_some({ bit_cast<u8*>(&event), sizeof(DeviceEvent) }));
```
Comment (kleinesfilmroellchen) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:196]: read_some may return before reading the entire buffer. Use read_until_filled

**Review by kleinesfilmroellchen — CHANGES_REQUESTED** [review]

File: Userland/Services/SystemServer/main.cpp [hunk 12]
```diff
@@ -392,49 +122,34 @@ static ErrorOr<void> prepare_bare_minimum_devtmpfs_directory_structure()
...
-    VERIFY(tty_group.has_value());
-    // FIXME: Try to find a way to not hardcode the major number of tty nodes.
-    TRY(chown_all_matching_device_nodes(tty_group.release_value(), 4));
-
-    auto audio_group = TRY(Core::System::getgrnam("audio"sv));
-    VERIFY(audio_group.has_value());
-    TRY(Core::System::chown("/dev/audio"sv, 0, audio_group->gr_gid));
-    TRY(chown_all_matching_device_nodes_under_specific_directory("/dev/audio"sv, audio_group.release_value()));
-    TRY(Core::System::endgrent());
+    TRY(Core::Process::spawn("/bin/DeviceMapper"sv, ReadonlySpan<StringView> {}, StringView {}, Core::Process::KeepAsChild::No));
```
Comment (kleinesfilmroellchen) [inline Userland/Services/SystemServer/main.cpp:144]: The explicit types in the two parameters shouldn’t be necessary.

**Review by kleinesfilmroellchen — APPROVED** [review]
Just a nitpick, let’s get this merged so we can move forward with USB device support

File: Userland/Services/SystemServer/main.cpp [hunk 13]
```diff
@@ -482,6 +482,28 @@ static ErrorOr<void> create_tmp_semaphore_directory()
...
+            g_services.append(move(service));
+        }
+    }
+    // After we've set them all up, activate them!
+    dbgln("Activating {} services...", g_services.size());
+    for (auto& service : g_services) {
+        dbgln_if(SYSTEMSERVER_DEBUG, "Activating {}", service->name());
+    }
+
+    for (auto& service : g_services) {
```
Comment (gmta) [inline Userland/Services/SystemServer/main.cpp:215]: Why are you looping over `g_services` twice? That `dbgln_if` above should probably be added here instead.

File: Userland/Services/SystemServer/main.cpp [hunk 14]
```diff
@@ -498,6 +520,42 @@ static ErrorOr<void> reopen_base_file_descriptors()
...
+            if (rc == 0) {
+                found_gpu_device = true;
+                break;
+            }
+            sleep(1);
+        }
+        if (!found_gpu_device) {
+            dbgln("WARNING: No device nodes at /dev/gpu/ directory after 10 seconds. This is probably a sign of disabled graphics functionality.");
+            dbgln("To cope with this, graphical mode will not be enabled.");
+            g_system_mode = "text";
```
Comment (gmta) [inline Userland/Services/SystemServer/main.cpp:254]: Instead of coming up with constants like these on the fly, we should probably have an enum or constexprs somewhere to reference.

File: Userland/Services/DeviceMapper/DeviceEventLoop.h [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    enum class MinorNumberAllocationType {
+        SequentialUnlimited,
+        SequentialLimited,
+    };
+
+    struct DeviceNodeMatch {
+        StringView permission_group;
+        StringView family_type_literal;
+        StringView path_pattern;
+        bool block_device;
```
Comment (gmta) [inline Userland/Services/DeviceMapper/DeviceEventLoop.h:30]: This could be an enum, similar to `MinorNumberAllocationType`

File: Userland/Services/DeviceMapper/main.cpp [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+ErrorOr<int> serenity_main(Main::Arguments)
+{
+    TRY(Core::System::unveil("/dev/"sv, "rwc"sv));
+    TRY(Core::System::unveil("/etc/group"sv, "rw"sv));
```
Comment (gmta) [inline Userland/Services/DeviceMapper/main.cpp:33]: Shouldn't you close with `unveil(nullptr, nullptr)` here?

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 17]
```diff
@@ -0,0 +1,241 @@
...
+static bool is_in_family_minor_number_range(DeviceNodeFamily const& family, MinorNumber minor_number)
+{
+    return family.base_minor_number() <= minor_number && static_cast<MinorNumber>(family.base_minor_number().value() + family.devices_symbol_suffix_allocation_map().size()) >= minor_number;
+}
+
+Optional<DeviceNodeFamily&> DeviceEventLoop::find_device_node_family(bool block_device, MajorNumber major_number, MinorNumber minor_number) const
+{
+    DeviceNodeFamily::Type type = block_device ? DeviceNodeFamily::Type::CharacterDevice : DeviceNodeFamily::Type::CharacterDevice;
+    for (auto const& family : m_device_node_families) {
+        if (family->major_number() == major_number && type == family->type() && is_in_family_minor_number_range(*family, minor_number))
```
Comment (gmta) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:65]: Code style: instead of `type == family->type()`, use `family->type() == type` so it's consistent with `family->major_number() == major_number`

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 18]
```diff
@@ -0,0 +1,241 @@
...
+    return node;
+}
+
+static ErrorOr<String> build_suffix_with_letters(size_t allocation_index)
+{
+    auto base_string = TRY(String::from_utf8(""sv));
+    do {
+        base_string = TRY(String::formatted("{:c}{}", 'a' + (allocation_index % 26), base_string));
+        allocation_index = allocation_index / 26;
+    } while (allocation_index > 0);
```
Comment (gmta) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:95]: This logic jumps from `z` to `ba` instead of `aa`, is that what you intended?

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 19]
```diff
@@ -0,0 +1,241 @@
...
+}
+
+static ErrorOr<void> prepare_permissions_after_populating_devtmpfs(StringView path, DeviceEventLoop::DeviceNodeMatch const& match)
+{
+    if (match.permission_group.is_null())
+        return {};
+    auto group = TRY(Core::System::getgrnam(match.permission_group));
+    VERIFY(group.has_value());
+    TRY(Core::System::chown(path, 0, group.value().gr_gid));
+    TRY(Core::System::endgrent());
```
Comment (gmta) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:111]: Apparently our LibC's `getgrnam` performs an automatic `endgrent` but `Core::System::getgrnam` does not; either we should align that behavior or this `endgrent` should take into account the possibility that `chown` fails and the group database isn't closed as a result of that.

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 20]
```diff
@@ -0,0 +1,241 @@
...
+        return Error::from_errno(ENODEV);
+    }
+    return {};
+}
+
+static ErrorOr<void> create_kcov_device_node()
+{
+    mode_t old_mask = umask(0);
+    TRY(Core::System::create_char_device("/dev/kcov"sv, 0666, 30, 0));
+    umask(old_mask);
```
Comment (gmta) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:191]: Maybe `ScopeGuard` the old umask instead?

File: Userland/Services/DeviceMapper/DeviceEventLoop.cpp [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+            TRY(register_new_device(event.is_block_device, event.major_number, event.minor_number));
+        } else if (event.state == DeviceEvent::State::Removed) {
+            if (event.major_number == 30 && event.minor_number == 0 && !event.is_block_device) {
+                dbgln("DeviceMapper: unregistering device failed: kcov tried to de-register itself!?");
+                continue;
+            }
+            if (auto error_or_void = unregister_device(event.is_block_device, event.major_number, event.minor_number); error_or_void.is_error())
+                dbgln("DeviceMapper: unregistering device failed: {}", error_or_void.error());
+        } else {
+            dbgln("DeviceMapper: Unhandled device event!");
```
Comment (gmta) [inline Userland/Services/DeviceMapper/DeviceEventLoop.cpp:235]: Suggestion: add `event.state` in the `dbgln` format so it's immediately clear which state triggered this.

**Review by ADKaster — APPROVED** [review]
This seems like an improvement to me! Good to see a nice way to block and wait for new devices and add them to /dev :)
